fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/properties/FDBPropertyKey.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * FDBPropertyKey.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb.properties;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.function.Supplier;
+
+@SuppressWarnings("unused")
+/**
+ * A key for a property that adopter configures for Record-Layer
+ * The value is lazily loaded from the supplier when the property is read
+ * If no supplier, a default value is used (this is not supposed to happen, since no reason that adopter actively configures a prop without providing a supplier)
+ */
+public final class FDBPropertyKey<T> {
+    @Nonnull
+    private final String name;
+    @Nonnull
+    private final T defaultValue;
+    @Nullable
+    private final Supplier<T> valueSupplier;
+
+    private FDBPropertyKey(@Nonnull final String name, @Nonnull T defaultValue, @Nullable final Supplier<T> valueSupplier) {
+        this.name = name;
+        this.defaultValue = defaultValue;
+        this.valueSupplier = valueSupplier;
+    }
+
+    @Nonnull
+    public String getName() {
```

Comment:
There needs to be documentation somewhere (potentially in the Javadoc comment here) about how the name for each property needs to be unique, and that we suggest using a reverse FQDN style naming convention to try and enforce that across packages.

Reply:
Added javadoc.